### PR TITLE
Adding Loop Fusion pass

### PR DESCRIPTION
**Pull request by shravankumar0811**
# **DO NOT FILE A PULL REQUEST**

This repository does not accept pull requests. Please follow http://llvm.org/docs/Contributing.html#how-to-submit-a-patch for contribution to LLVM.

# **DO NOT FILE A PULL REQUEST**

### CODE REVIEW

File: llvm/lib/Transforms/LoopFusion/LoopFusion.cpp [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+    SmallVector<Loop *> LoopVector;
+    LoopInfo *LI = &getAnalysis<LoopInfoWrapperPass>().getLoopInfo();
+
+    for (auto *L : *LI) {
+      LoopVector.push_back(L);
+    }
+
+    // Function to perform basic checks on the two loops
+    if (fuseCheck(LoopVector[1], LoopVector[0]))
```
Comment (theguydk) [inline llvm/lib/Transforms/LoopFusion/LoopFusion.cpp]: This is going to give out of bound when there is only one loop in the test. Add proper checks before calling the functions

File: llvm/lib/Transforms/LoopFusion/LoopFusion.cpp [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+  }
+
+  void fuseBody(Loop *Loop1, Loop *Loop2, Function &F) {
+    BasicBlock *Body1 = nullptr;
+    BasicBlock *Body2 = nullptr;
+    BasicBlock *Header1 = nullptr;
+    BasicBlock *Latch1 = nullptr;
+    BasicBlock *Exit2 = nullptr;
+
+    Body1 = getBody(Loop1);
```
Comment (theguydk) [inline llvm/lib/Transforms/LoopFusion/LoopFusion.cpp]: use assertion/check wherever necessary.

getBody() here might return nullptr

File: llvm/lib/Transforms/LoopFusion/LoopFusion.cpp (hunk 1 above)
Comment (prasadkpbhat) [inline llvm/lib/Transforms/LoopFusion/LoopFusion.cpp]: do not hardcode.This condition causes error if func has single loop.
How u r assuming only 1st and 2nd loop are valid candidates for fusion?
U need to search all loops within a function.

File: llvm/lib/Transforms/LoopFusion/LoopFusion.cpp [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    LoopInfo *LI = &getAnalysis<LoopInfoWrapperPass>().getLoopInfo();
+
+    for (auto *L : *LI) {
+      LoopVector.push_back(L);
+    }
+
+    // Function to perform basic checks on the two loops
+    if (fuseCheck(LoopVector[1], LoopVector[0]))
+      // Function to perform fusing on the two loops
+      fuseBody(LoopVector[1], LoopVector[0], F);
```
Comment (prasadkpbhat) [inline llvm/lib/Transforms/LoopFusion/LoopFusion.cpp]: same here also

File: llvm/lib/Transforms/LoopFusion/LoopFusion.cpp [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    BasicBlock *Bb1 = Loop1->getExitBlock();
+    BasicBlock *Bb2 = Loop2->getLoopPreheader();
+
+    //  If exit block and preHeader are not same.
+    if (Bb1 != Bb2) {
+      if (Bb1->size() != 1)
+        return false;
+      if (Bb1->getTerminator()->getSuccessor(0) != Bb2)
+        return false;
+      if (Bb1 == nullptr || Bb2 == nullptr) {
```
Comment (prasadkpbhat) [inline llvm/lib/Transforms/LoopFusion/LoopFusion.cpp]: Null check should be at the beginning.

File: llvm/lib/Transforms/LoopFusion/LoopFusion.cpp [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+  BasicBlock *getLoopExit(Loop *L) { return L->getExitBlock(); }
+
+  bool adjacent(Loop *Loop1, Loop *Loop2) {
+
+    BasicBlock *Bb1 = Loop1->getExitBlock();
+    BasicBlock *Bb2 = Loop2->getLoopPreheader();
+
+    //  If exit block and preHeader are not same.
+    if (Bb1 != Bb2) {
+      if (Bb1->size() != 1)
```
Comment (prasadkpbhat) [inline llvm/lib/Transforms/LoopFusion/LoopFusion.cpp]: what is the need of these many if statements?as for all if u r returning false and in the end also u r returning false.
Simply u can do like this right?
if (Bb1 != Bb2) {
   return false
}

Comment (prasadkpbhat) [reply llvm/lib/Transforms/LoopFusion/LoopFusion.cpp]: return Bb1 == Bb2

File: llvm/lib/Transforms/LoopFusion/LoopFusion.cpp [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+    // Check if the limit integer is same.
+    if (limitValue(L1) != limitValue(L2)) {
+      llvm::errs() << "The loop check limiting value is not same.CANNOT fuse\n";
+      return false;
+    }
+    return true;
+  }
+
+  // Check if the start value is same.
+  int startValue(Loop &LoopV, ScalarEvolution &SE) {
```
Comment (prasadkpbhat) [inline llvm/lib/Transforms/LoopFusion/LoopFusion.cpp]: What if start value is runtime variable?
return value should be of Value type

Comment (prasadkpbhat) [reply llvm/lib/Transforms/LoopFusion/LoopFusion.cpp]: what is the use of SCEV here?

File: llvm/lib/Transforms/LoopFusion/LoopFusion.cpp [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    for (auto &IndVar : LoopV.getHeader()->phis()) {
+      Value *V = IndVar.getOperand(1);
+      auto startValue = dyn_cast<ConstantInt>(V);
+      return startValue->getSExtValue();
```
Comment (prasadkpbhat) [inline llvm/lib/Transforms/LoopFusion/LoopFusion.cpp]: Add null check here

Comment (prasadkpbhat) [inline llvm/lib/Transforms/LoopFusion/LoopFusion.cpp]: Dont assume start value to be of int type

File: llvm/lib/Transforms/LoopFusion/LoopFusion.cpp [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    }
+    return {};
+  }
+
+  // Check if the limit value is same.
+  Value *limitValue(Loop *LoopV) {
+    Value *end;
+    for (Use &U : LoopV->getHeader()->getFirstNonPHI()->operands()) {
+      if (!dyn_cast<PHINode>(U.get())) {
+        Instruction *I = dyn_cast<Instruction>(U.get());
```
Comment (prasadkpbhat) [inline llvm/lib/Transforms/LoopFusion/LoopFusion.cpp]: U need to search for icmp instruction in header.
From there u can get upper bound

File: llvm/lib/Transforms/LoopFusion/LoopFusion.cpp [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+      }
+    }
+    return {};
+  }
+
+  // Function to get Loop Header Blocks.
+  BasicBlock *getHeader(Loop *L) { return L->getHeader(); }
+
+  // Function to get Loop Latch Blocks.
+  BasicBlock *getLoopLatch(Loop *L) {
```
Comment (prasadkpbhat) [inline llvm/lib/Transforms/LoopFusion/LoopFusion.cpp]: u can directly use getLoopLatch from LoopInfo pass

File: llvm/lib/Transforms/LoopFusion/LoopFusion.cpp [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+    Exit2 = getLoopExit(Loop2);
+
+    PHINode *Phi1 = Loop1->getCanonicalInductionVariable();
+    PHINode *Phi2 = Loop2->getCanonicalInductionVariable();
+
+    // Replace the use of induction variable of 2nd loop with that of 1st loop.
+    Phi2->replaceAllUsesWith(Phi1);
+
+    for (BasicBlock &BB : F) {
+      BranchInst *BI = dyn_cast<BranchInst>(BB.getTerminator());
```
Comment (prasadkpbhat) [inline llvm/lib/Transforms/LoopFusion/LoopFusion.cpp:100]: add null check

File: llvm/lib/Transforms/LoopFusion/LoopFusion.cpp [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+    BasicBlock *Body2 = nullptr;
+    BasicBlock *Header1 = nullptr;
+    BasicBlock *Latch1 = nullptr;
+    BasicBlock *Exit2 = nullptr;
+
+    Body1 = getBody(Loop1);
+    Body2 = getBody(Loop2);
+    Header1 = getHeader(Loop1);
+    Latch1 = getLoopLatch(Loop1);
+    Exit2 = getLoopExit(Loop2);
```
Comment (SHBharadwaj) [inline llvm/lib/Transforms/LoopFusion/LoopFusion.cpp]: directly use Loop2->getExitBlock()

File: llvm/lib/Transforms/LoopFusion/LoopFusion.cpp [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+  void fuseBody(Loop *Loop1, Loop *Loop2, Function &F) {
+    BasicBlock *Body1 = nullptr;
+    BasicBlock *Body2 = nullptr;
+    BasicBlock *Header1 = nullptr;
+    BasicBlock *Latch1 = nullptr;
+    BasicBlock *Exit2 = nullptr;
+
+    Body1 = getBody(Loop1);
+    Body2 = getBody(Loop2);
+    Header1 = getHeader(Loop1);
```
Comment (SHBharadwaj) [inline llvm/lib/Transforms/LoopFusion/LoopFusion.cpp]: use Loop1->getHeader()

File: tests/loop_fuse/loop_fuse_out.ll [hunk 13]
```diff
@@ -0,0 +1,75 @@
+; ModuleID = 'loop_fuse.ll'
```
Comment (SHBharadwaj) [inline tests/loop_fuse/loop_fuse_out.ll:1]: Am I missing something here ? Shouldn't two loops be fused here ?

File: llvm/lib/Transforms/LoopFusion/LoopFusion.cpp [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+
+    // Function to perform basic checks on the two loops
+    if (fuseCheck(LoopVector[1], LoopVector[0]))
+      // Function to perform fusing on the two loops
+      fuseBody(LoopVector[1], LoopVector[0], F);
+
+    return false;
+  }
+
+  void fuseBody(Loop *Loop1, Loop *Loop2, Function &F) {
```
Comment (SHBharadwaj) [inline llvm/lib/Transforms/LoopFusion/LoopFusion.cpp:75]: Please add a detailed comment as to what is the transformation happening under this & the logic used.

File: llvm/lib/Transforms/LoopFusion/LoopFusion.cpp [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+
+  // Function to get Loop Body Blocks.
+  BasicBlock *getBody(Loop *L) {
+    for (BasicBlock *BB : L->getBlocks()) {
+      BasicBlock *HeaderBlock = L->getHeader();
+      if ((HeaderBlock != BB) && !(L->isLoopLatch(BB))) {
+        return BB;
+      }
+    }
+    return {};
```
Comment (SHBharadwaj) [inline llvm/lib/Transforms/LoopFusion/LoopFusion.cpp]: return nullptr;

**Review by shravan-kumar-0811 — APPROVED** [review]

**Pull request merged**